model/flow/service_event.go

```diff
@@ -61,69 +69,95 @@ func (sel ServiceEventList) EqualTo(other ServiceEventList) (bool, error) {
 	return true, nil
 }
 
+// ServiceEventMarshaller marshals and unmarshals all types of service events.
 type ServiceEventMarshaller interface {
-	Unmarshal(b []byte) (ServiceEvent, error)
-	UnmarshalWithType(
-		b []byte,
-		eventType ServiceEventType,
-	) (
-		ServiceEvent,
-		error,
-	)
+	// UnmarshalWrapped unmarshals the service event and returns it as a wrapped ServiceEvent type.
+	// The input bytes must be encoded as a generic wrapped ServiceEvent type.
+	// Forwards errors from the underlying marshaller (treat errors as you would from eg. json.Unmarshal)
+	UnmarshalWrapped(b []byte) (ServiceEvent, error)
+	// UnmarshalWithType unmarshals the service event and returns it as a wrapped ServiceEvent type.
+	// The input bytes must be encoded as a specific event type (for example, EpochSetup).
+	// Forwards errors from the underlying marshaller (treat errors as you would from eg. json.Unmarshal)
+	UnmarshalWithType(b []byte, eventType ServiceEventType) (ServiceEvent, error)
 }
 
 type marshallerImpl struct {
-	MarshalFunc   func(v interface{}) ([]byte, error)
-	UnmarshalFunc func(data []byte, v interface{}) error
+	marshalFunc   func(v interface{}) ([]byte, error)
+	unmarshalFunc func(data []byte, v interface{}) error
 }
 
+var _ ServiceEventMarshaller = (*marshallerImpl)(nil)
+
 var (
+	// CAUTION: Json and MsgPack are to be used only for trusted data sources
 	ServiceEventJSONMarshaller = marshallerImpl{
-		MarshalFunc:   json.Marshal,
-		UnmarshalFunc: json.Unmarshal,
+		marshalFunc:   json.Marshal,
+		unmarshalFunc: json.Unmarshal,
 	}
+	// CAUTION: Json and MsgPack are to be used only for trusted data sources
 	ServiceEventMSGPACKMarshaller = marshallerImpl{
-		MarshalFunc:   msgpack.Marshal,
-		UnmarshalFunc: msgpack.Unmarshal,
+		marshalFunc:   msgpack.Marshal,
+		unmarshalFunc: msgpack.Unmarshal,
 	}
 	ServiceEventCBORMarshaller = marshallerImpl{
-		MarshalFunc:   cborcodec.EncMode.Marshal,
-		UnmarshalFunc: cbor.Unmarshal,
+		marshalFunc:   cborcodec.EncMode.Marshal,
+		unmarshalFunc: cbor.Unmarshal,
 	}
 )
 
-func (marshaller marshallerImpl) Unmarshal(b []byte) (
-	ServiceEvent,
-	error,
-) {
-	var enc map[string]interface{}
-	err := marshaller.UnmarshalFunc(b, &enc)
+// UnmarshalWrapped unmarshals the service event and returns it as a wrapped ServiceEvent type.
+// The input bytes must be encoded as a generic wrapped ServiceEvent type.
+// Forwards errors from the underlying marshaller (treat errors as you would from eg. json.Unmarshal)
+func (marshaller marshallerImpl) UnmarshalWrapped(b []byte) (ServiceEvent, error) {
+	var eventTypeWrapper struct {
+		Type ServiceEventType
+	}
+	err := marshaller.unmarshalFunc(b, &eventTypeWrapper)
```

Comment:
That is certainly the default behaviour. It seems like [CBOR includes an option](https://github.com/fxamacker/cbor/blob/master/decode.go#L513) to fail decoding if it finds a field which doesn't match the struct it is marshalling into. Our use of the decoder doesn't enable that option. That will be true for all CBOR encoding in Flow. 

We could accommodate that change by changing the service event structure to something like:
```go
type ServiceEvent struct {
  Type string
  Event []byte
}
```
and encoding outer then inner. Probably other approaches too, like optional fields for each possible event type. 

My feeling is this would be a moderately large refactor, because some components depend on the ability to both operate generically on `ServiceEvent` instances and on the underlying specific type.

Reply:
- this issue allows a surface area for spam -- sending additional garbage data in a network message with impact:
  - use extra bandwidth without detection
  - use extra memory without detection

action:
- create issues:
  - enable strict mode in global cbor decoder
  - after above, modify service event decoding to be compatible with global strict CBOR decoder

Reply:
We can also change to the structure you mentioned:
```
type ServiceEvent struct {
  Type string
  Event []byte
}
```
And use local strict CBOR encoder which will deal with encoding both `ServiceEvent` and underlying type in `Event`. If some fields are optional usually that can be encoded using special comments.

Reply:
Issues:
- https://github.com/onflow/flow-go/issues/5534
- https://github.com/onflow/flow-go/issues/5535